### ISSUE TITLE
Don't use LocalDateTime for java8 datetime mapping

### ISSUE THREAD

**Issue by cbornet** [swagger-api/swagger-codegen#2138]
LocalDateTime doesn't have any time zone so it's error prone for systems that don't communicate in the same timezone.
Instant, ZonedDateTime or OffsetDateTime types should be considered instead.

**Comment by wing328** [comment]
FYI. C# already supports `OffsetDateTime` via CLI option

**Comment by cbornet** [comment]
Also by default, when Jackson serializes/deserializes a LocalDateTime, it uses "yyyy-MM-dd'T'HH:mm:ss.SSS" without the 'Z' or the time offset. So it is not compatible with swagger's datetime format.

**Comment by ebelanger** [comment]
#3002

I am leaving a link to this pull request and the alternative solution so that other may benefit like I did.

**Comment by cbornet** [comment]
I used Instant because:
* ZonedDateTime has time-zone info (eg. Europe/Paris) which is not supported by rfc3339
* Older implementations of date-time (eg. in java 6) may only use rfc822 time zones (-0800 instead of -08:00) so I generally prefer to force UTC at the API level. Feel free to comment if you think OffsetDateTime would be better.

**Comment by cbornet** [comment]
Just saw an issue with using ```Instant```: we will probably not be able to deserialize rfc3339 dates with time offset (at least that's the case with Jackson).

And the issue with rfc3339 date offset not being supported is also on android implementation of [SimpleDateFormat](https://developer.android.com/reference/java/text/SimpleDateFormat.html) (which is more problematic than java6 in the end).
This makes me think that some of our java client implementations may have issues on android (eg. java jersey client uses the 'XXX' SimpleDateFormat to be compliant with rfc3339 but that will probably fail on android).

So to sum up:
* ```Instant``` will not deserialize rfc3339 but will serialize correctly for android (I don't know about other clients)
* ```OffsetDateTime``` will deserialize correctly but the serialized will probably not be readable by android by default and maybe other clients. Note that we can probably adapt the serializers to force the output in UTC/Zulu (it's possible with Jackson).

So I think we use OffsetDateTime and configure the serializers in the templates to output in Zulu. Ok with that ?

**Comment by rsltrifork** [comment]
We prefer Instant over OffsetDateTime because forcing UTC on the API level is a good thing.
Is there any way to make swagger generate timestamps as java.time.Instant instead of OffsetDateTime?

**Comment by cbornet** [comment]
The problem is that jackson will not deserialize date-times to Instant if thy have an offset.
It is preferable to configure jackson to serialize OffsetDateTime in UTC. This can be done by adding a custom serializer to the JavaTimeModule.

**Comment by cbornet** [comment]
If you really want to use Instant then use ```--type-mappings DateTime=Instant --import-mappings Instant=java.time.Instant```

**Comment by tsujeeth** [comment]
> "--type-mappings DateTime=Instant --import-mappings Instant=java.time.Instant"

@cbornet Do you have an example of doing the same in `swagger-codegen-maven-plugin`?

**Comment by tsujeeth** [comment]
FWIW, the example below helps in setting date-time's type to java.time.Instant instead of OffsetDateTime.

```xml
<execution>
    <id>swagger-path-specs.definitions</id>
    <goals>
        <goal>generate</goal>
    </goals>
    <configuration>
        <verbose>true</verbose>
        <inputSpec>${basedir}/src/main/resources/specs/swagger.yaml</inputSpec>
        <language>java</language>
        <configOptions>     
            <sourceFolder>src/gen/java/main</sourceFolder>
            <useBeanValidation>true</useBeanValidation>
            <dateLibrary>java8</dateLibrary>
        </configOptions>    
        <typeMappings>      
            <typeMapping>OffsetDateTime=Instant</typeMapping>
        </typeMappings>     
        <importMappings>    
            <importMapping>java.time.OffsetDateTime=java.time.Instant</importMapping>
        </importMappings>   
        <library>jersey2</library>
    </configuration>    
</execution>
```

**Comment by rkcreadev** [comment]
@tsujeeth thx! It's work for. But for my case (when I using LocalDateTime) I use:

> <execution>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                        <configuration>
                            <inputSpec>${swagger.specification.path}</inputSpec>
                            <language>spring</language>
                            <library>spring-boot</library>
                            <generateSupportingFiles>false</generateSupportingFiles>
                            <generateApiDocumentation>true</generateApiDocumentation>
                            <modelPackage>${swagger.model.package}</modelPackage>
                            <apiPackage>${swagger.controller.package}</apiPackage>
                            <configOptions>
                                
                                <interfaceOnly>true</interfaceOnly>
                                <delegatePattern>true</delegatePattern>
                                <java8>true</java8>
                            </configOptions>
                            <typeMappings>
                                <typeMapping>OffsetDateTime=LocalDateTime</typeMapping>
                            </typeMappings>
                            <importMappings>
                                <importMapping>java.time.OffsetDateTime=java.time.LocalDateTime</importMapping>
                            </importMappings>
                        </configuration>
                    </execution>
